### PR TITLE
bio field now part of User; status moved off of DrupalUser

### PR DESCRIPTION
**Pull request by jywarren**
This also establishes a 'token' field for API use and deprecates the `location_privacy` field of User.

* [x] all tests pass -- `rake test:all`
* [x] code is in uniquely-named feature branch, and has been rebased on top of latest master (especially if you've been asked to make additional changes)
* [x] pull request is descriptively named with #number reference back to original issue

@ryzokuken - here I'm initializing a `token` field. You can autogen random tokens in this PR if you like; add to the migration file in each user. We only have to generate them for users of non-zero status.

### CODE REVIEW

File: db/migrate/20170704165711_add_user_bio_and_token.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  def up
+    add_column :rusers, :bio, :text
+    add_column :rusers, :token, :string
+    add_column :rusers, :status, :integer, default: 0
+    remove_column :rusers, :location_privacy
+
+    # copy bios into new fields for non-spam users
+    DrupalUsers.where('status != 0').each do |u|
+      user = u.user
+      user.status = u.status
```
Comment (jywarren) [inline db/migrate/20170704165711_add_user_bio_and_token.rb]: Right here you could do something like `user.token = SecureRandom.hex`

Comment (ryzokuken) [inline db/migrate/20170704165711_add_user_bio_and_token.rb]: @jywarren okay, I am setting up plots2 again and will do this hopefully by today.

Comment (ryzokuken) [inline db/migrate/20170704165711_add_user_bio_and_token.rb]: @jywarren I don't think `SecureRandom.hex` will ensure uniqueness, though. What do you say?

**Comment by jywarren** [comment]
You can concatenate the UID to the beginning?

**Comment by ryzokuken** [comment]
@jywarren `SecureRandom.uuid` sounds better. Although I would like to point out:

> This is because the probability that the 16-byte or 128-bit token is nonunique is so vanishingly small that it is virtually zero. There is only a 50% chance of there being any repetitions after approximately 264 = 18,446,744,073,709,551,616 = 1.845 x 1019 tokens have been generated. If you start generating one billion tokens per second, it will take approximately 264/(109*3600*24*365.25) = 600 years until there is a 50% chance of there having occurred any repetitions at all.

**Comment by jywarren** [comment]
That's fine. We need to regenerate the schema too, and perhaps remove a few unused tables like searches, location tags

**Comment by ryzokuken** [comment]
Basically, these are our options:

```ruby
SecureRandom.uuid           # => "8efb5d40-32d8-43c8-a92d-d5f048c8729c"
SecureRandom.hex            # => "bad6650fc0142451e624bd89b6aa3acf"
SecureRandom.urlsafe_base64 # => "vp6xjBgi48Jgag6dqH8niw"
```

**Comment by jywarren** [comment]
I trust your judgement on this!

**Comment by ryzokuken** [comment]
@jywarren `SecureRandom.uuid` looks a little more verbose, but I will use it. Will make the required changes as soon as the environment is setup locally. A few thing I would like to confirm though:

1. How do I add to this PR? Do I copy your branch, make changes and make a PR for your fork's branch? When you push my changes, they will show up in here, I guess.

2. Apart from adding `user.token = SecureRandom.uuid` in https://github.com/publiclab/plots2/pull/1497/files/4014dfc777c83f9f97491676b9ffd0aadf2250c4#diff-85d8a967f5e9f12cfb43f842e560d394, what all other tasks do I need to perform? Which specific tables need to be removed?

**Comment by jywarren** [comment]
I can remove the tables, that's not a big deal, or if you like you can
drop: `searches` and `location_tags`.

You should be able to open a PR against the feature branch I'm using, and I
can merge it into my feature branch :-) then the PRs combine!

**Comment by ryzokuken** [comment]
@jywarren Awesome. I will look more into safely dropping tables from the schema while `bundle install` executes. On another note, are we using any dependencies we could avoid? The CI build takes way too much time to finish.

**Comment by jywarren** [comment]
You can drop tables in the migration itself and then copy your new
generated local schema over the schema example file.

@icarito was talking about caching some parts to save time. I also think we
could move some of our tests from integration into lower level unit tests
(and likewise actual code) as unit tests are more self contained and less
context dependent, and so run much faster. You could open an issue around
optimizing tests!

**Comment by ryzokuken** [comment]
Definitely. One of the new contributors was also talking about Makefiles which reminded of the PR I once made for the same. I will consult @icarito and make one for real this time.

**Comment by ryzokuken** [comment]
@jywarren we figured out a workaround for `Gem::LoadError: mysql2 is not part of the bundle. Add it to your Gemfile.`, right? I am getting this error :/

**Comment by jywarren** [comment]
I think you figured this out? Sorry i wasn't able to respond faster!

**Comment by ryzokuken** [comment]
@jywarren Yes, you already fixed it, I was using an older version :P

I will submit a PR for this one in a moment, thanks for your patience.

**Comment by ryzokuken** [comment]
@jywarren Made a PR. Please look into it. Awaiting your approval before going ahead with the migration.

**Comment by ryzokuken** [comment]
@jywarren

```
ActiveRecord::StatementInvalid: Mysql2::Error: BLOB/TEXT column 'params' used in key specification without a key length: CREATE  INDEX `poly_params_request_index` ON `impressions` (`impressionable_type`, `impressionable_id`, `params`)

/app/db/schema.rb:193:in `block in <top (required)>'

/app/db/schema.rb:14:in `<top (required)>'

Mysql2::Error: BLOB/TEXT column 'params' used in key specification without a key length

/app/db/schema.rb:193:in `block in <top (required)>'

/app/db/schema.rb:14:in `<top (required)>'

Tasks: TOP => db:schema:load
```

File: db/migrate/20170704165711_add_user_bio_and_token.rb (hunk 1 above)
Comment (jywarren) [inline db/migrate/20170704165711_add_user_bio_and_token.rb]: We need to regenerate the schema file. Doing this now!

Comment (ryzokuken) [inline db/migrate/20170704165711_add_user_bio_and_token.rb]: @jywarren great.

**Comment by ryzokuken** [comment]
@jywarren There's a pending migration. Do you want me to make a PR with the migration done and schema replaced?

**Comment by jywarren** [comment]
Sure, or I can do it in an hour. Just forgot a step.

**Comment by ryzokuken** [comment]
@jywarren No problem. I'll do it right now, all I need to do is `rake db:migrate` and then copy my schema to the example schema, right?

**Comment by jywarren** [comment]
Huh, did u notice there were changes to a few other tables in your schema? I think they're SQLite MySQL differences. Could you remove them? Thanks!

**Comment by ryzokuken** [comment]
@jywarren I will. Just a minute.

**Comment by ryzokuken** [comment]
@jywarren were they still there? (the sqlite differences)

Strange. I made sure to manually remove them all.

**Comment by ryzokuken** [comment]
@jywarren Nope, here's my `schema.rb.example`:

```ruby
add_index "impressions", ["controller_name", "action_name", "ip_address"], :name => "controlleraction_ip_index"
  add_index "impressions", ["controller_name", "action_name", "request_hash"], :name => "controlleraction_request_index"
  add_index "impressions", ["controller_name", "action_name", "session_hash"], :name => "controlleraction_session_index"
  add_index "impressions", ["impressionable_type", "impressionable_id", "ip_address"], :name => "poly_ip_index"
  add_index "impressions", ["impressionable_type", "impressionable_id", "params"], :name => "poly_params_request_index", :unique => false, :length => { "params" => 255 }
  add_index "impressions", ["impressionable_type", "impressionable_id", "request_hash"], :name => "poly_request_index"
  add_index "impressions", ["impressionable_type", "impressionable_id", "session_hash"], :name => "poly_session_index"
  add_index "impressions", ["impressionable_type", "message", "impressionable_id"], :name => "impressionable_type_message_index", :unique => false, :length => { "message" => 255 }
  add_index "impressions", ["user_id"], :name => "index_impressions_on_user_id"
```

File: db/schema.rb.example [hunk 2]
```diff
@@ -190,26 +190,12 @@ ActiveRecord::Schema.define(:version => 20170208174046) do
   add_index "impressions", ["controller_name", "action_name", "request_hash"], :name => "controlleraction_request_index"
   add_index "impressions", ["controller_name", "action_name", "session_hash"], :name => "controlleraction_session_index"
   add_index "impressions", ["impressionable_type", "impressionable_id", "ip_address"], :name => "poly_ip_index"
-  add_index "impressions", ["impressionable_type", "impressionable_id", "params"], :name => "poly_params_request_index", :unique => false, :length => { "params" => 255 }
+  add_index "impressions", ["impressionable_type", "impressionable_id", "params"], :name => "poly_params_request_index"
```
Comment (ryzokuken) [inline db/schema.rb.example]: This is what went wrong, but this one was done in another commit, I think.

File: db/schema.rb.example [hunk 3]
```diff
@@ -190,26 +190,12 @@ ActiveRecord::Schema.define(:version => 20170208174046) do
...
   add_index "impressions", ["impressionable_type", "impressionable_id", "request_hash"], :name => "poly_request_index"
   add_index "impressions", ["impressionable_type", "impressionable_id", "session_hash"], :name => "poly_session_index"
-  add_index "impressions", ["impressionable_type", "message", "impressionable_id"], :name => "impressionable_type_message_index", :unique => false, :length => { "message" => 255 }
+  add_index "impressions", ["impressionable_type", "message", "impressionable_id"], :name => "impressionable_type_message_index"
```
Comment (ryzokuken) [inline db/schema.rb.example]: This too.

**Comment by jywarren** [comment]
Can you try reverting those lines? Thanks!

**Comment by ryzokuken** [comment]
@jywarren sure, I will.

**Comment by jywarren** [comment]
aha! now this is my error. looking now.

**Comment by ryzokuken** [comment]
@jywarren But we did replace `schema.rb.example` didn't we?

**Comment by ryzokuken** [comment]
@jywarren the tests need to be changed to match the current schema.

File: db/migrate/20170704165711_add_user_bio_and_token.rb [hunk 4]
```diff
@@ -0,0 +1,26 @@
+class AddUserBioAndToken < ActiveRecord::Migration
+  def up
+    add_column :rusers, :bio, :text, limit: 2147483647, default: ''
```
Comment (jywarren) [inline db/migrate/20170704165711_add_user_bio_and_token.rb]: it's odd; the error I'm seeing is:

```
Error: test_confirm_user_reset_key_not_visible_on_profile_to_non-admins(UsersControllerTest): ActionView::Template::Error: wrong argument type nil (expected String)
app/views/users/profile.html.erb:149:in `to_html'
```

that's on this line:

```ruby
<p><small><%= raw auto_link(RDiscount.new(@profile_user.bio).to_html, :sanitize => false) %></small></p>
```

And `@profile_user` is a `User` as of this line: https://github.com/jywarren/plots2/blob/user-bio-and-token/app/controllers/users_controller.rb#L107

So, given this default, it should have a value of '', not `nil`. What's the deal?

Comment (jywarren) [inline db/migrate/20170704165711_add_user_bio_and_token.rb]: I'm adding a unit test to ensure that a new `User` record has a non-nil `bio`.

Comment (ryzokuken) [inline db/migrate/20170704165711_add_user_bio_and_token.rb]: @jywarren I'll try to look into it.

Comment (ryzokuken) [inline db/migrate/20170704165711_add_user_bio_and_token.rb]: Oh, okay. Yes, in the migration we set the default bio to `''`, right?

Comment (jywarren) [inline db/migrate/20170704165711_add_user_bio_and_token.rb]: yes, that's what's confusing!

Comment (jywarren) [inline db/migrate/20170704165711_add_user_bio_and_token.rb]: OK, so the '' is not working:

```
Failure: <nil> expected to not be nil.
test_user_creation(UserTest)
test/unit/user_test.rb:17:in `block in <class:UserTest>'
```

Comment (ryzokuken) [inline db/migrate/20170704165711_add_user_bio_and_token.rb]: @jywarren Why not? It should have when I ran the migration...

Comment (jywarren) [inline db/migrate/20170704165711_add_user_bio_and_token.rb]: Apparently default values are not universally supported??? whoa. I'll try setting it manually.

Comment (ryzokuken) [inline db/migrate/20170704165711_add_user_bio_and_token.rb]: Using `rails console`?

Comment (ryzokuken) [inline db/migrate/20170704165711_add_user_bio_and_token.rb]: @jywarren Oh, the fixtures weren't updated?

Comment (jywarren) [inline db/migrate/20170704165711_add_user_bio_and_token.rb]: ugh yeah. i should've done this all at once, locally, but i kept thinking it was a small fix.

Comment (ryzokuken) [inline db/migrate/20170704165711_add_user_bio_and_token.rb]: @jywarren I'm glad we have this sorted. What next for this PR?

**Comment by jywarren** [comment]
This is good to go! If you can now open a new PR with the comment_controller.rb work to enable posting a comment using only a token, and not just if you're logged in, that'd be great.

**Comment by ryzokuken** [comment]
@jywarren I would. I think I should go to bed now but I definitely will do so tomorrow morning.

**Comment by jywarren** [comment]
good night! Thanks for your hard work as always!
